pkg/envoy/xds/ack.go

```diff
@@ -210,7 +253,53 @@ func (m *AckingResourceMutatorWrapper) HandleResourceVersionAck(version uint64,
 
 		// Completion didn't match or is still waiting for some ACKs. Keep it
 		// in the pending list.
-		remainingCompletions = append(remainingCompletions, comp)
+		remainingCompletions[comp.completion] = comp
+	}
+
+	m.pendingCompletions = remainingCompletions
+}
+
+func (m *AckingResourceMutatorWrapper) HandleResourceVersionNack(version uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string) {
+	ackLog := log.WithFields(logrus.Fields{
+		logfields.XDSVersionInfo: version,
+		logfields.XDSClientNode:  node,
+		logfields.XDSTypeURL:     typeURL,
+	})
+
+	nodeID, err := m.nodeToID(node)
+	if err != nil {
+		// Ignore NACKs from unknown or misconfigured nodes which have invalid
+		// node identifiers.
+		ackLog.Warning("invalid ID in Node identifier; ignoring NACK")
+		return
+	}
+
+	m.locker.Lock()
+	defer m.locker.Unlock()
+
+	remainingCompletions := make(map[*completion.Completion]*pendingCompletion, len(m.pendingCompletions))
+
+	for _, comp := range m.pendingCompletions {
```

Comment:
You can now use the key of that:
`for comp, pending := range m.pendingCompletions {`

Reply:
Done.